### PR TITLE
Refactor trackViews() to reliably remove views

### LINKED ISSUE #4064: Code views are not cleaned up anymore
Some ViewResolvers return an `element` for the CodeView that does not match the element found with the `selector`. We currently store CodeView state in a Map keyed by element, so this means we never clean them up.

It seems like we cannot make the element always match what's found with the selector. Since we cannot run the ViewResolver for removed elements (they look at outside state), the only solution is using a `Set` instead and checking if  `removedNode.contains(codeView.element)` for every CodeView.

See discussion in https://github.com/sourcegraph/sourcegraph/pull/3976

### PR DESCRIPTION
**Pull request by lguychard**
Fixes #4064

Refactors `trackViews()` to stop relying on a `ViewResolver`'s selector to detect
removed views. We now inspect `removedNodes` to see if any of the removed nodes
is a known view, or contains one or more known views.

With this refactor, the `viewStates` logic is now encapsulated within `trackViews()`,
which simply emits every added views as a `ViewWithSubscriptions`, and takes care of calling
`view.subscriptions.unsubscribe()` when the view is removed from the page. This prevents
the consumer from having to reimplement the `viewStates` pattern.

Test plan: The view addition/removal logic is well tested by existing unit tests, but performance should be benchmarked as view removal is now more expensive.

TODO:
- [x] Fix remaining failing test `handleContentViews() › detects addition, mutation, and removal of content views`
- [x] Consider consolidating `code_views.test.ts` with `views.test.ts`

### CODE REVIEW

**Comment by lguychard** [comment]
I've looked at some profiles and interestingly, this performs consistently _better_ than the status quo.

Test protocol:
- Start on https://github.com/sourcegraph/sourcegraph/pull/2344/files, make sure all 75 code views are added
- Click to navigate to the "Conversation" tab
- Profile from mouseup to the end of the following animation frame

Baseline: no Sourcegraph extension activated:

image

Sourcegraph dev extension built from `master`:

image

Sourcegraph dev extension built from this branch:

image

Another takeaway is that GitHub performance for this user action (switching between tabs on a large PR) is bad to start with, and we make it at least 2x worse 😑

File: browser/src/libs/code_intelligence/code_intelligence.tsx [hunk 1]
```diff
@@ -477,213 +472,224 @@ export function handleCodeHost({
         observeOn(animationFrameScheduler)
     )
 
-    interface CodeViewState {
-        subscriptions: Subscription
-        roots: WorkspaceRootWithMetadata[]
+    const rootRefCounts = new Map<string, number>()
```
Comment (felixfbecker) [inline browser/src/libs/code_intelligence/code_intelligence.tsx]: Always document the semantics of primitive type arguments to collections

```suggestion
    /** Map from workspace URI to number of editors referencing it */
    const rootRefCounts = new Map<string, number>()
```

File: browser/src/libs/code_intelligence/code_intelligence.tsx [hunk 2]
```diff
@@ -477,213 +472,224 @@ export function handleCodeHost({
...
+                isActive: true,
+            }
+            const editorId = extensionsController.services.editor.addEditor(editorData)
+            const scope: PartialCodeEditor = {
+                ...editorData,
+                ...editorId,
+                model,
+            }
+            const rootURI = toRootURI(fileInfo)
+            addRootRef(rootURI, fileInfo.rev || '')
```
Comment (felixfbecker) [inline browser/src/libs/code_intelligence/code_intelligence.tsx]: Empty string seems wrong to me, should be `undefined` (which both `fileInfo.rev` and `WorkspaceRootWithMetadata.inputRev` allow
```suggestion
            addRootRef(rootURI, fileInfo.rev)
```

File: browser/src/libs/code_intelligence/code_intelligence.tsx [hunk 3]
```diff
@@ -477,213 +472,224 @@ export function handleCodeHost({
...
-                            commitID: fileInfo.baseCommitID,
-                        }),
-                        inputRevision: fileInfo.baseRev || '',
-                    })
-                }
+                const baseRootURI = toRootURI({
+                    repoName: fileInfo.baseRepoName,
+                    commitID: fileInfo.baseCommitID,
+                })
+                addRootRef(baseRootURI, fileInfo.baseRev || '')
```
Comment (felixfbecker) [inline browser/src/libs/code_intelligence/code_intelligence.tsx]: Same here
```suggestion
                addRootRef(baseRootURI, fileInfo.baseRev)
```

File: browser/src/libs/code_intelligence/code_views.test.ts [hunk 4]
```diff
@@ -38,7 +38,7 @@ describe('code_views', () => {
                     toArray()
                 )
                 .toPromise()
-            expect(detected).toEqual([{ ...codeViewSpec, element, type: 'added' }])
+            expect(detected.map(({ subscriptions, ...rest }) => ({ ...rest }))).toEqual([{ ...codeViewSpec, element }])
```
Comment (felixfbecker) [inline browser/src/libs/code_intelligence/code_views.test.ts]:
```suggestion
            expect(detected.map(({ subscriptions, ...rest }) => rest)).toEqual([{ ...codeViewSpec, element }])
```

File: browser/src/libs/code_intelligence/code_views.test.ts [hunk 5]
```diff
@@ -149,26 +127,19 @@ describe('code_views', () => {
...
+                [{ addedNodes: [document.body], removedNodes: [] }],
+                [{ addedNodes: [], removedNodes: [container] }],
+            ])
+            await mutations
                 .pipe(
                     trackCodeViews({
                         codeViewResolvers: [toCodeViewResolver(selector, codeViewSpec)],
-                    })
+                    }),
+                    bufferCount(1),
```
Comment (felixfbecker) [inline browser/src/libs/code_intelligence/code_views.test.ts]: What's the point of a `bufferCount`, if the buffer is just 1?

File: browser/src/libs/code_intelligence/views.ts [hunk 6]
```diff
@@ -22,78 +33,72 @@ export interface ViewResolver<V> {
...
+                    if (!(node instanceof HTMLElement)) {
+                        continue
+                    }
+                    const view = viewStates.get(node)
+                    if (view) {
+                        view.subscriptions.unsubscribe()
+                        viewStates.delete(node)
+                        continue
+                    }
+                    for (const viewElement of viewStates.keys()) {
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/code_intelligence/views.ts]: No reason, fixed

File: browser/src/libs/code_intelligence/code_views.test.ts (hunk 5 above)
Comment (lguychard) [reply browser/src/libs/code_intelligence/code_views.test.ts]: No reason, this was copy-pasta from another test where the `bufferCount` actually served a purpose.

File: browser/src/libs/code_intelligence/content_views.ts [hunk 7]
```diff
@@ -40,85 +40,59 @@ export function handleContentViews(
...
-
-                if (contentViewEvent.type === 'removed' || contentViewEvent.type === 'updated') {
-                    const contentViewState = contentViewStates.get(contentViewEvent.element)
-                    if (contentViewState) {
-                        contentViewState.subscriptions.unsubscribe()
-                        contentViewStates.delete(contentViewEvent.element)
-                    }
-                }
+                console.log(`Content view ${contentViewEvent.type}`, { contentViewEvent })
+                contentViewEvent.subscriptions.add(() => console.log(`Content view removed`, { contentViewEvent }))
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/code_intelligence/content_views.ts]: They're for production, mimicking [the logs we already had](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/browser/src/libs/code_intelligence/code_intelligence.tsx#L489:25).

File: browser/src/libs/code_intelligence/code_intelligence.tsx [hunk 8]
```diff
@@ -453,21 +451,18 @@ export function handleCodeHost({
...
-                return codeViewEvent.resolveFileInfo(codeViewEvent.element, platformContext.requestGraphQL).pipe(
-                    mergeMap(fileInfo =>
-                        fetchFileContents(fileInfo, platformContext.requestGraphQL).pipe(
-                            map(fileInfoWithContents => ({
-                                fileInfo: fileInfoWithContents,
-                                ...codeViewEvent,
-                            }))
-                        )
+        mergeMap(codeViewEvent =>
+            codeViewEvent.resolveFileInfo(codeViewEvent.element, platformContext.requestGraphQL).pipe(
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/code_intelligence/code_intelligence.tsx]: `trackViews()` no longer emits on removed views. It emits every added view as a `ViewWithSubscriptions`, with `view.subscriptions` being unsubscribed from when the view is removed.

This is essentially an encapsulation within `trackViews()` of the view states pattern, which was implemented by every `trackViews()` consumer before:
- [Code views](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/browser/src/libs/code_intelligence/code_intelligence.tsx#L480-485)
- [Content views](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/browser/src/libs/code_intelligence/content_views.ts#L43-47)
- [Text fields](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/browser/src/libs/code_intelligence/text_fields.tsx#L37-41)

File: browser/src/libs/code_intelligence/code_views.test.ts [hunk 9]
```diff
@@ -1,15 +1,19 @@
-import { of, Subject } from 'rxjs'
+import { of, Subscription } from 'rxjs'
```
Comment (lguychard) [inline browser/src/libs/code_intelligence/code_views.test.ts]: This file now only contains tests that are specific to code views: resolving from spec vs. removing from selector.

File: browser/src/libs/code_intelligence/views.test.ts [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+                trackViews([{ selector: '.view', resolveView: element => ({ element }) }]),
+                bufferCount(3),
+                switchMap(views =>
+                    Promise.all(views.map(view => new Promise(resolve => view.subscriptions.add(resolve))))
+                )
+            )
+            .toPromise()
+    })
+
+    test('removes a view without depending on its resolver', async () => {
```
Comment (lguychard) [inline browser/src/libs/code_intelligence/views.test.ts]: @felixfbecker this test verifies that the issues described in https://github.com/sourcegraph/sourcegraph/pull/3976#issuecomment-492186685 are solved.

**Pull request merged**